### ISSUE TITLE
What's the point of `Denomination::MilliSatoshi`?

### ISSUE THREAD

**Issue by TheBlueMatt** [rust-bitcoin/rust-bitcoin#2820]
Its incredibly confusing to have a `MilliSatoshi` `Denomination`, implying `Amount`s might support it, but instead if you try to use it you get an error. It seems like it just sets the wrong impression but doesn't really add any value since you can't really safely use it anywhere? ISTM it should be dropped.

**Comment by tcharding** [comment]
Oh, I always assumed it was used by lightning. If LDK doesn't use it then why did we add it originally?

**Comment by apoelstra** [comment]
Lol, I also assumed that somehow it was used by the LN people. I'm happy to drop it.

**Comment by tcharding** [comment]
So do we want an `Amount` that fully supports LDK or are you guys resigned to keeping your own?

**Comment by tcharding** [comment]
The inner `u64` inside `Amount` is private so we can put what ever we want in there, for example (excluding the obvious perf hit) times it by 1000 and make it represent milistats. To save us hanging ourselves we'd probably want a `Satoshi` type and not use `u64`s ever.

**Comment by jirijakes** [comment]
By the way, fedimint also defines [its own type](https://github.com/fedimint/fedimint/blob/3a08b3b2319be0460bbde07b82410cd61ba537e1/fedimint-core/src/lib.rs#L168):

```rs
/// Represents an amount of BTC inside the system. The base denomination is
/// milli satoshi for now, this is also why the amount type from rust-bitcoin
/// isn't used instead.
pub struct Amount {
    pub msats: u64,
}
```

**Comment by apoelstra** [comment]
>I sincerely hope that no one is doing enough amount construction in an inner loop that this is noticeable :).

@yancyribbons had this situation once running some sort of exhaustive coin selection algo.

@tcharding : My viev is:

* `Amount` should represent an on-chain amount. No millisats. If we tried to support millisats in `Amount` we would run into error cases when we tried to put this object on-chain, as Matt says.
* We could additionally support some sort of `LightningAmount` which probably we'd just copy straight from ldk. But it may make more sense for users who want Lightning amounts to just use ldk.
* Fedimint's notion of sub-satoshi amounts is, I suspect, different from that of Lightning. Maybe they have the same range and in theory they could share a type, but I'm not sure that will always be the case, given the iteration that Fedimint is doing on amount privacy, denominations, etc. So while I appreciate the desire for "the whole ecosystem to have one `Amount` type" I don't really think this is possible.

**Comment by tcharding** [comment]
Ouch, that hurts. @dpc and @TheBlueMatt you guys are long in the tooth enough to know that upstreaming is the best way to reduce your workload - the `bitcoin::Amount` hands down should support LDK and Fedimint.

Since we only just did the last release there is no rush but here is the issue: https://github.com/rust-bitcoin/rust-bitcoin/issues/2824

**Comment by tcharding** [comment]
Cool, thanks. I consider myself schooled. I didn't see the above message before opening an issue (now closed), or chiding @dcp and @TheBlueMatt about not upstreaming, I've deleted it, my sincere apologies.

**Comment by tcharding** [comment]
> Amount should represent an on-chain amount. No millisats. If we tried to support millisats in Amount we would run into error cases when we tried to put this object on-chain, as Matt says.

"on chain" just means what we spit out of `consensus_encode`, right? That is not tied to the internal representation of `Amount`.

The privacy stuff in Fedimint is in the future so I don't have an opinion on that. I believe we can technically support LDK, Fedimint, and on-chain users with a single `Amount` type. I see no reason that if msats are a thing then that the `Amount` type should not be able to support them.

**Comment by apoelstra** [comment]
> "on chain" just means what we spit out of `consensus_encode`, right? That is not tied to the internal representation of `Amount`.

Right.

> The privacy stuff in Fedimint is in the future so I don't have an opinion on that. I believe we can technically support LDK, Fedimint, and on-chain users with a single `Amount` type. I see no reason that if msats are a thing then that the `Amount` type should not be able to support them.

Vacuously agreed. msats are not a thing :) There is no such thing as a millisatoshi except as an accounting fiction in non-bitcoin protocols.

**Comment by tcharding** [comment]
> msats are not a thing

Ok, if that is our position then this issue is aptly named. Shall we just remove all mention of it?

**Comment by tcharding** [comment]
> ISTM it should be dropped.

lol, @TheBlueMatt said exactly that right at the start - don't mind me ...

**Comment by dpc** [comment]
> Fedimint's notion of sub-satoshi amounts is, I suspect, different from that of Lightning. 

Fedimint's `Amount`s are the way they are, precisely to be compatible with LN. I was actually going to propose `LightnintAmount` after getting to keyboard, but I see it was already mentioned. It would most probably work for Fedimint. And even if it doesn't, I think it makes sense to have both "on chain amount" and "ln amount" as interoperable concepts, and projects can always have `From` back-and-forth for interoperability.

Though I must say having a custom `Amount` in a project is not a big deal really anyway.


Ping @elsirion for more Fedimint PoV on the matter.

**Comment by elsirion** [comment]
> * Fedimint's notion of sub-satoshi amounts is, I suspect, different from that of Lightning.

It's the same as lightning. As @dpc mentions LN's support for it is the reason for Fedimint support.

> … given the iteration that Fedimint is doing on amount privacy, denominations, etc.

Optionally blinded amounts will likely result in us creating an enum over either some blinded amount commitment or a plain amount, both being msat denominated.

**Comment by dpc** [comment]
> I think its rather universally accepted at this point that msats were a mistake in LN and we shouldn't have done them. I'd encourage fedimint to drop them if at all possible at this point :)

Hopefully not to derail the discussion, but any good authoritative link to read more about it? I heard it before, but never investigated it. In Fedimint msats seem entirely practical without many (any?) downsides. And for stuff like micropayments etc. could work great. So I wonder if the problems that made it a mistake in LN, do actually apply to ECash.

**Comment by Kixunil** [comment]
The ligthning amount already exists in the `ln_types` crate. Perhaps you guys could finally start using it. :) Ping me if you want access permissions/transfer/whatever. I don't have much time lately but I'm slowly going back. And BTW I'll be at BTC Prague in case someone is attending too.

Also, @TheBlueMatt FYI msat is parsed correctly if you have multiples of thousand. This is intentional but I'm not claiming it to be practical. I guess displaying sats as msats is a more sensible use case.

**Comment by yancyribbens** [comment]
> I sincerely hope that no one is doing enough amount construction in an inner loop that this is noticeable :).

> @yancyribbons had this situation once running some sort of exhaustive coin selection algo.

The rust BnB coinselection algo I wrote suffered performance problems (when compared to bitcoin cores BnB selection) when using `Amount` types.  However I tracked the issue to the use of `checked` arithmetic which was default.  It didn't have anything to do with construction of an amount type (done outside of the hot path).

TLDR the `unchecked` arithmetic variants on Amount fixed the perf problems I was experiencing with coin-selection.

@apoelstra thanks for the mention, however you made I very minor spelling mistake with your mention :).  Just happened to notice this conversation and was reading through it out of interest.
